src/modm/platform/i2c/stm32-extended/i2c_timing_calculator.hpp

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2018, Christopher Durand
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_STM32_I2C_TIMING_CALCULATOR_HPP
+#define MODM_STM32_I2C_TIMING_CALCULATOR_HPP
+
+#include <tuple>
+#include <optional>
+#include <array>
+#include <cmath>
+
+namespace modm
+{
+
+namespace platform
+{
+
+struct I2cMasterTimings
+{
+	uint8_t prescaler;
+	uint8_t sdaDel;
+	uint8_t sclDel;
+	uint8_t sclLow;
+	uint8_t sclHigh;
+};
+
+struct I2cParameters
+{
+	uint32_t peripheralClock;
+	uint32_t targetSpeed;
+	uint16_t tolerance;
+	uint8_t digitalFilterLength;
+	bool enableAnalogFilter;
+	float riseTime;
+	float fallTime;
+};
+
+/**
+ * STM32 extended I2C timing calculator
+ *
+ * @author		Christopher Durand
+ * @ingroup		modm_platform_i2c
+ */
+class I2cTimingCalculator
+{
+public:
+	constexpr I2cTimingCalculator(const I2cParameters& parameters) : params{parameters} {}
+
+	constexpr std::optional<I2cMasterTimings>
+	calculateTimings() const
+	{
+		uint16_t prescalerMask{findValidPrescalers()};
+		if(prescalerMask == 0) {
+			return std::nullopt;
+		}
+
+		bool found = false;
+		float lastError = 1e10;
+		uint8_t bestSclLow = 0;
+		uint8_t bestSclHigh = 0;
+		uint8_t bestPrescaler = 16;
+
+		for(int prescaler = 15; prescaler >= 0; --prescaler)
+		{
+			if(not (prescalerMask & (1 << prescaler))) {
+				continue;
+			}
+
+			for(uint16_t sclLow = 0; sclLow <= 255; ++sclLow) {
+				for(uint16_t sclHigh = 0; sclHigh <= 255; ++sclHigh) {
+					std::optional<float> speed = calculateSpeed(prescaler, sclLow, sclHigh);
```

Comment:
Can you guesstimate a sclLow/sclHigh combination and search upwards or downwards from there? Are there any bounds on the error? Is there a gradient of the error, you can then use for bisection?